server/src/weaverbird/backends/mongo_translator/steps/date_extract.py

```diff
@@ -0,0 +1,320 @@
+            12,
+            {'$subtract': [{'$month': f'${step.column}'}, 1]},
+        ],
+    }
+
+
+def _extract_previous_week(step: DateExtractStep) -> MongoStep:
+    return {
+        # We subtract to the target date 7 days in milliseconds
+        '$week': {'$subtract': [f'${step.column}', 7 * 24 * 60 * 60 * 1000]},
+    }
+
+
+def _extract_previous_quarter(step: DateExtractStep) -> MongoStep:
+    return {
+        '$switch': {
+            'branches': [
+                {'case': {'$lte': [{'$divide': [{'$month': f'${step.column}'}, 3]}, 1]}, 'then': 4},
+                {'case': {'$lte': [{'$divide': [{'$month': f'${step.column}'}, 3]}, 2]}, 'then': 1},
+                {'case': {'$lte': [{'$divide': [{'$month': f'${step.column}'}, 3]}, 3]}, 'then': 2},
+            ],
+            'default': 3,
+        },
+    }
+
+
+def _extract_previous_iso_week(step: DateExtractStep) -> MongoStep:
+    return {
+        # We subtract to the target date 7 days in milliseconds
+        '$isoWeek': {'$subtract': [f'${step.column}', 7 * 24 * 60 * 60 * 1000]},
+    }
+
+
+_ADVANCED_DATE_EXTRACT_MAP = {
+    'quarter': _extract_quarter,
+    'firstDayOfYear': _extract_first_day_of_year,
+    'firstDayOfMonth': _extract_first_day_of_month,
+    'firstDayOfWeek': _extract_first_day_of_week,
+    'firstDayOfQuarter': _extract_first_day_of_quarter,
+    'firstDayOfIsoWeek': _extract_first_day_of_iso_week,
+    'previousDay': _extract_previous_day,
+    'firstDayOfPreviousYear': _extract_first_day_of_previous_year,
+    'firstDayOfPreviousMonth': _extract_first_day_of_previous_month,
+    'firstDayOfPreviousWeek': _extract_first_day_of_previous_week,
+    'firstDayOfPreviousQuarter': _extract_first_day_of_previous_quarter,
+    'firstDayOfPreviousIsoWeek': _extract_first_day_of_previous_iso_week,
+    'previousYear': _extract_previous_year,
+    'previousMonth': _extract_previous_month,
+    'previousWeek': _extract_previous_week,
+    'previousQuarter': _extract_previous_quarter,
+    'previousIsoWeek': _extract_previous_iso_week,
+}
+
+_DATE_EXTRACT_MAP = {
+    'year': '$year',
+    'month': '$month',
+    'day': '$dayOfMonth',
+    'week': '$week',
+    'dayOfYear': '$dayOfYear',
+    'dayOfWeek': '$dayOfWeek',
+    'isoYear': '$isoWeekYear',
+    'isoWeek': '$isoWeek',
+    'isoDayOfWeek': '$isoDayOfWeek',
+    'hour': '$hour',
+    'minutes': '$minute',
+    'seconds': '$second',
+    'milliseconds': '$millisecond',
+}
+
+
+def translate_date_extract(step: DateExtractStep) -> List[MongoStep]:
+    new_columns = []
+    add_fields = {}
+
+    date_info: list
+    # For retrocompatibility
+    if step.operation:
+        date_info = [step.operation] if step.operation else step.date_info
+        new_columns = [
+            step.new_column_name if step.new_column_name else f"{step.column}_{step.operation}"
+        ]
+    else:
+        date_info = step.date_info.copy()
+        new_columns = deepcopy(step.new_columns)
```

Comment:
Not a blocking point, but just for curiosity, why not using the `.copy()` also here instead of `deepcopy` ? 🤔

Reply:
Because I thought that `NewColumn` was an object but it's only an alias for `str`. You're right, I'll update it :+1: